### PR TITLE
envoy: Wait for endpoint restoration before listening on xDS socket

### PR DESCRIPTION
**Pull request by jrajahalme**
Delay listening on the Cilium xDS socket until endpoints have been restored. This eliminates the problem where Envoy times out on the initial fetch on the xDS streams.

```release-note
Cilium agent now waits until endpoints have restored before starting accepting new xDS streams.
```

### CODE REVIEW

**Review by sayboras — APPROVED** [review]

**Comment by jrajahalme** [comment]
Rebased for a CI fix

**Review by tommyp1ckles — APPROVED** [review]

File: pkg/envoy/grpc.go [hunk 1]
```diff
@@ -50,14 +50,30 @@ func (s *xdsServer) startXDSGRPCServer(listener net.Listener, config map[string]
 
 	reflection.Register(grpcServer)
 
+	ctx, cancel := context.WithCancel(context.Background())
 	go func() {
+		if s.restorerPromise != nil {
+			restorer, err := s.restorerPromise.Await(ctx)
+			if err == nil && restorer != nil {
+				log.Infof("Envoy: Waiting for endpoint restoration before serving xDS resources...")
+				err = restorer.WaitForEndpointRestore(ctx)
```
Comment (tommyp1ckles) [inline pkg/envoy/grpc.go:59]: Quick question, to confirm - is this err value that is intended to be checked below against `context.Canceled`?

**Pull request merged**